File: src/cirrus-ngs/server/WGSPipeline/src/WGSPipelineLauncher/src/org/ucsd/ccbb/PBSTracker.java
package org.ucsd.ccbb;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.sql.Timestamp;
import java.util.Date;

/**
 * The class is to track the PBS queue status. 
 * For example, how many jobs are running in the queue.
 * 
 * @author Guorong Xu
 */
public class PBSTracker {
	public void trackPBSQueue(int minutes, String shellScript)
	{
		int jobs = 0;
	    while (true) {
	    	try {
	    			Date date= new Date();
	    			Thread.sleep(minutes * 60000);
	                
		        	String[] commandLine = new String[]{"qstat"};
		        	Process process = Runtime.getRuntime().exec(commandLine);

	                String line;
	                BufferedReader in = new BufferedReader(new InputStreamReader(process.getInputStream()));
	                
	                boolean isDone = true;
	                int count = 0;
	                while ((line = in.readLine()) != null) {
	                    if (line.indexOf(shellScript) > -1)
	                    		isDone = false;
	                    count++;
	                }
	                //ignore the first two lines 
	                if (count > 2)
	                {
	                	if ((count - 2) != jobs)
	                	{
	                		jobs = count - 2;
	                		System.out.println("");
	                		//System.out.println(new Timestamp(date.getTime()) + "  job(s) are running...");
	                		System.out.println(new Timestamp(date.getTime()) + " running job(s): " + (count - 2));
	                	}
	                	//else
	                		//need to be modified here: print out the current time
	                		//how to print out how long the program has run?
	                		//System.out.print("...");
	                }
	                
	                if (isDone)
	                {
	                	System.out.println("");
	                	System.out.println(new Timestamp(date.getTime()) + " No jobs is running...");
	                	break;
	                }
	                
	                in.close();
	    	} catch (Exception e) {
	    	}
	    }
	}
	   
	/**
	 * Checks how many command lines started with the word "node" ?
	 * @return	an integer indicating the number of commands started with the 
	 * 			word "node"
	 */
	public int checkNodeNum()
	{
		int num = 0;
	    try {
	    		String[] commandLine = new String[]{"qhost"};
	    	   	Process process = Runtime.getRuntime().exec(commandLine);
			
	    	   	String line;
	    	   	BufferedReader in = new BufferedReader(new InputStreamReader(process.getInputStream()));
				    
	    	   	while ((line = in.readLine()) != null) {
	    			Date date= new Date();
	    	   		System.out.println(new Timestamp(date.getTime()) + line);
	    	   		if (line.startsWith("node"))
	    	   			num++;
				}
	            in.close();
	    } catch (Exception e) {
			Date date2= new Date();
			System.out.println(new Timestamp(date2.getTime()));
	        System.out.println(e);
	    }
	       
	    return num;
	}
}

File: src/cirrus_ngs/server/deprecated/SmallRNASeqPipeline/PBSTracker.py
__author__ = 'Guorong Xu<g1xu@ucsd.edu>'

import re
import time
import subprocess

def trackPBSQueue(minutes, shell_script):

    jobs = 0
    while True:
        time.sleep(minutes * 60)

        is_done = True
        index = 0

        cmd = 'qstat'
        stdout = subprocess.check_output([cmd])

        lines = re.split(r'\n+', stdout)

        for line in lines:
            if line.find(shell_script) > -1 and not isDone(line):
                is_done = False
                index = index + 1

        if index > 0:
            if index != jobs:
                jobs = index
                print ("")
                print (str(index) , " job(s) are running...")

        if is_done:
            print ("")
            print ("No jobs are running...")
            break

def isDone(qstat_job):
    status_list = qstat_job.strip().split()
    if len(status_list) == 1:
        return True

    qstat_job_status = status_list[4]

    if re.match("d.*", qstat_job_status):
        return True
    return False

if __name__ == "__main__":
    trackPBSQueue(1, "PBS_Tracker")
